src/battleSystem/hpGauge.mjs

```diff
@@ -3,119 +3,343 @@
+const full = 0;
+const damaged = 1;
+const empty = 2;
+
+/* to-do:
+1. implement a setter for capacity
+	- currently editing capacity after creating the bar will break it
+	- setter would need to
+		a) create new reserves if needed
+		b) edit this.fraction
+		c) edit this._capacity - all refs to this.capacity would change to that
+2. refactor/centralise more of the maths
+3. double check all colour references are correct
+*/
 
 export default
 class HPGauge
 {
-	constructor(capacity, sectorSize = 100, color = Color.White, maxSectors = 'auto')
+	constructor(capacity, sectorSize = 100, color = white, maxSectors = 0)
 	{
-		this.borderColor = Color.Black.fadeTo(color.a);
+		// basic details
 		this.capacity = capacity;
+		this.sectorSize = sectorSize; //hp in a full bar
+		this.numReserves = Math.ceil(capacity / sectorSize) - 1;
+		this.maxSectors = maxSectors; //note 0 is used for infinity
+		this.reserves = [];
+
+		//house keeping
+		this.numCombosRunning = 0;
+		this.newReading = capacity;
+		this.oldReading = capacity;
+		let remainder = capacity % sectorSize;
+		this.fraction = remainder  === 0 ? 1 : remainder / sectorSize;
+		this.fading = false;
+		this.drawBackground = true;
+
+		//stats on what is shown
+		this._reading = 0; // the total HP remaining
+		this.damage = 0;
+		this._damageShown = 0; //total value of damage shown
+		this._numReservesFilled = this.numReserves; //filled reserves
+		this._numReservesDamaged = 0; //damaged i.e. draining away reserves
+		this._numReservesEmpty = 0; //balancing figure but still necessary...
+
+		//sizes
+		this._x = 0;
+		this._y = 0;
+		this._height = 0;
+		this._width = 0;
+		this._slotYSize = 0;
+		this._slotXSize = 0;
+
+		// shaders
+		this.shader = new Shader({ //shader for most things
+			fragmentFile: '#/shaders/image.frag.glsl',
+			vertexFile:   '#/shaders/image.vert.glsl',
+		});
+		this.hpShader = this.shader.clone(); //hp bar shader used for colours (and fading) of hp bar
+		this.backgroundShader = this.shader.clone(); // background shader for the faint background bar
+
+		//timers
 		this.colorFadeDuration = 0;
 		this.colorFadeTimer = 0;
-		this.damage = 0;
-		this.damageColor = Color.DarkRed.fadeTo(color.a);
-		this.damageFadeness = 1.0;
 		this.drainSpeed = 1.0;
-		this.emptyColor = Color.of('#202020').fadeTo(color.a);
+		this.drainTimer = 0;
 		this.fadeSpeed = 0.0;
-		this.fadeness = 1.0;
-		this.hpColor = color.clone();
-		this.isVisible = true;
-		this.maxSectors = maxSectors;
-		this.newColor = color.clone();
-		this.newReading = this.capacity;
-		this.numCombosRunning = 0;
-		this.oldColor = color.clone();
-		this.oldReading = this.capacity;
-		this.reading = this.capacity;
-		this.sectorSize = sectorSize;
+		this._fadeness = 1.0;
+		this._damageFadeness = 1;
+
+		//colour objects
+		this._hpColor = null;
+		this.newColor = null;
+		this.oldColor = null;
+		this.tintColor = {
+			r : color.r,
+			g : color.g,
+			b : color.b,
+			a : color.a
+		};//intentionally make a copy
```

Comment:
I assume this is a performance trick to avoid calling `color.r` etc. getters?  I guess it avoids a native function call... but how often are we accessing those in practice? (I'm mostly concerned about the extra verbosity over just calling `color.clone()`)

Reply:
When fading in and out and when changing colour these get used 2-3 times a frame for the duration of the fade/change - there's no benefit to having a colour object as only going to supply them to the shader as a float vector.

8 lines of extra code to save a large quantity of CPU instructions for each fade.

Reply:
Also allowed me to avoid a new Colour call each frame whilst tween'ing fades.